src/io/image_olm_parser.rs

```diff
@@ -71,57 +73,39 @@ pub fn render(
 }
 
 // TODO: handle unwrap of image::open properly
-pub fn parse(filename: &str, chunk_size: usize) -> (Rules, PixelKeys, MSu16xNU, Vec<Chunk>) {
+pub fn parse(filename: &str, chunk_size: usize) -> (Rules, PixelKeys, MSu16xNU, IndexMap<Chunk, u16>) {
     let img = image::open(filename).unwrap().to_rgb8();
     let pixel_aliases = alias_pixels(&img);
-    let chunk_frequencies = chunk_image(img, chunk_size, &pixel_aliases, false);
-    // convert frequencies into a list of unique chunks
-    // TODO: could use chunk keys directly in future
-    let mut chunks: Vec<Chunk> = chunk_frequencies
-        .keys()
-        .fold(Vec::new(), |mut acc, chunk| {
-            acc.push(chunk.clone());
-            acc
-        });
-    let overlap_rules = overlaps(&chunks, chunk_size);
+    let chunk_frequencies = chunk_image(img, chunk_size, &pixel_aliases, true, false);
+    let overlap_rules = overlaps(&chunk_frequencies, chunk_size);
 
-    if chunks.len() > MSu16xNU::len() {
-        println!("Chunks LEN: {}", chunks.len());
+    if chunk_frequencies.len() > MSu16xNU::len() {
+        println!("Chunks LEN: {}", chunk_frequencies.len());
         panic!("labels multiset not large enough to store all unique chunks")
     }
 
-    // put frequencies into multi set
-    let mut all_labels = chunks
-        .iter()
-        .enumerate()
-        .fold(MSu16xNU::empty(), |mut acc, (index, chunk)| {
-            let frequency = chunk_frequencies.get(chunk).unwrap();
-            acc.insert(index, *frequency as u16);
-            // acc.insert(index, 1); // previous implementation
-            acc
-        });
-
+    let all_labels = chunk_frequencies.values().collect();
     let raw_graph = create_raw_graph(&all_labels, chunk_size, (3, 3));
     let mut pruned_rules: Rules = HashMap::new();
 
-    (0..all_labels.count_non_zero())
+    (0..MSu16xNU::len())
```

Comment:
Wouldn't this result in too many iterations?

Reply:
Correct good spot, though it doesn't seem to affect anything being this way. None the less will revert.